lib/charms/lte_core_interface/v0/lte_core_interface.py

```diff
@@ -0,0 +1,246 @@
+
+    on = CoreRequirerCharmEvents()
+
+    def __init__(self, charm: CharmBase, relationship_name: str):
+        """Init."""
+        super().__init__(charm, relationship_name)
+        self.charm = charm
+        self.relationship_name = relationship_name
+        self.framework.observe(
+            charm.on[relationship_name].relation_changed, self._on_relation_changed
+        )
+
+    @staticmethod
+    def _relation_data_is_valid(remote_app_relation_data: dict) -> bool:
+        try:
+            validate(instance=remote_app_relation_data, schema=REQUIRER_JSON_SCHEMA)
+            return True
+        except exceptions.ValidationError as e:
+            logger.error("Invalid relation data: %s", e)
+            return False
+
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handler triggered on relation changed event.
+
+        Args:
+            event: Juju event (RelationChangedEvent)
+
+        Returns:
+            None
+        """
+        relation = self.model.get_relation(self.relationship_name)
+        if not relation:
+            logger.warning("No relation with name %s", self.relationship_name)
+            return
+        if not relation.app:
+            logger.warning("No remote application in relation: %s", self.relationship_name)
+            return
+        remote_app_relation_data = relation.data[relation.app]
+        if not self._relation_data_is_valid(dict(remote_app_relation_data)):
+            logger.warning(
+                "Provider relation data did no pass JSON Schema validation: \n%s",
+                event.relation.data[event.app],  # type: ignore[index]
+            )
+            return
+        self.on.lte_core_available.emit(
+            mme_ipv4_address=remote_app_relation_data["mme_ipv4_address"],
+        )
+
+
+class CoreProvides(Object):
+    """Class to be instantiated by the charm providing the core."""
+
+    def __init__(self, charm: CharmBase, relationship_name: str):
+        """Init."""
+        super().__init__(charm, relationship_name)
+        self.relationship_name = relationship_name
+        self.charm = charm
+
+    @staticmethod
+    def mme_ipv4_address_is_valid(mme_ipv4_address: str) -> bool:
+        """Returns whether mme ipv4 address is valid."""
+        # TODO: Is this approach ok?
+        try:
+            IPv4Address(mme_ipv4_address)
+            return True
+        except:  # noqa: E722
+            return False
+
+    def set_core_information(self, mme_ipv4_address: str):
+        """Sets mme_ipv4_address in the application relation data.
+
+        Args:
+            mme_ipv4_address: MME ipv4 address
+        Returns:b
+            None
+        """
+        if not self.charm.unit.is_leader():
+            raise RuntimeError("Unit must be leader to set application relation data.")
+        if not self.mme_ipv4_address_is_valid(mme_ipv4_address):
+            raise AddressValueError("Invalid MME IPv4 address.")
```

Comment:
I believe libs should communicate with the charm using events rather than exceptions. In this case, you could emit an event and use it to block the `Provider` charm. 
What do you think?

Reply:
Why do you say this?
> I believe libs should communicate with the charm using events rather than exceptions.
But seems like a good idea

Reply:
> Why do you say this?

That's what I've learnt while working on the interface in the Alertmanager.

Reply:
I agree, I had similar experience, let the library put data into the databag, read it from there and emit custom events, then leave any other responsibility to the charm.

Reply:
After discussing with Guillaume, we'll keep it as is. But thanks for the comment anyways, definitely something to keep in mind.